From b5d889b1cb66f572a0f43ba4d26423208f0d698c Mon Sep 17 00:00:00 2001
From: Ramasamy Kaliappan <rkaliapp@qti.qualcomm.com>
Date: Tue, 19 Aug 2025 20:39:07 +0530
Subject: [PATCH] QSDK:PROTOCOL: Hostapd: Hidden BSSID Support for Tx and
 Non-Tx BSS

Hidden BSSID profiles will be included in beacon and probe responses.
The SSID value will either be omitted or set to all zeros, based on
the configured value of ignore_broadcast_ssid.

Currently, Few issues have been identified where the SSID is not
included in unicast probe responses for requests targeting a hidden,
non-transmitted BSS, and where sending a probe request to a
non-transmitted BSS may unintentionally expose the SSID if the
transmitted BSS is hidden.

Add support to include the SSID in unicast probe responses for such
requests, depending on the ignore_broadcast_ssid configuration for
transmitted or non-transmitted BSS profiles

Fixes: 27faa86e745db (hostapd: QSDK patches update for package upgrade 13.0)

Co-developed-by: Sowmiya Sree Elavalagan <ssreeela@qti.qualcomm.com>
Signed-off-by: Sowmiya Sree Elavalagan <ssreeela@qti.qualcomm.com>
Signed-off-by: Ramasamy Kaliappan <rkaliapp@qti.qualcomm.com>
---
 src/ap/beacon.c     | 47 ++++++++++++++++++++++++++++++++++++---------
 src/ap/ieee802_11.c |  8 ++++++++
 2 files changed, 46 insertions(+), 9 deletions(-)

--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -876,14 +876,26 @@ static u8 * hostapd_probe_resp_fill_elem
 	u8 p_ext_cap = 0;
 	bool bcast_prb_resp = false;
 
+	hapd = hostapd_mbssid_get_tx_bss(hapd);
+
 	epos = pos + len;
 
 	*pos++ = WLAN_EID_SSID;
-	*pos++ = hapd->conf->ssid.ssid_len;
-	os_memcpy(pos, hapd->conf->ssid.ssid,
-		  hapd->conf->ssid.ssid_len);
-	pos += hapd->conf->ssid.ssid_len;
 
+	if (hapd->conf->ignore_broadcast_ssid && hapd != hapd_probed) {
+		if (hapd->conf->ignore_broadcast_ssid == 2) {
+			*pos++ = hapd->conf->ssid.ssid_len;
+			os_memset(pos, 0, hapd->conf->ssid.ssid_len);
+			pos += hapd->conf->ssid.ssid_len;
+		} else {
+			*pos++ = 0; /* empty SSID */
+		}
+	} else {
+		*pos++ = hapd->conf->ssid.ssid_len;
+		os_memcpy(pos, hapd->conf->ssid.ssid,
+				hapd->conf->ssid.ssid_len);
+		pos += hapd->conf->ssid.ssid_len;
+	}
 	/* Supported rates */
 	pos = hostapd_eid_supp_rates(hapd, pos);
 
@@ -1178,7 +1190,7 @@ static void hostapd_gen_probe_resp(struc
 	params->resp->u.probe_resp.capab_info |=
 		host_to_le16(hostapd_critical_update_capab(hapd));
 
-	pos = hostapd_probe_resp_fill_elems(hapd, params,
+	pos = hostapd_probe_resp_fill_elems(hapd_probed, params,
 					    params->resp->u.probe_resp.variable,
 					    buflen);
 
